### PR TITLE
fix: Fix the Panic in recover()

### LINKED ISSUE #7139: Malformed workflow causes workflow-controller to crash loop
## Summary

What happened/what you expected to happen?

Submitting a new workflow with a malformed status causes the workflow controller to crash loop, preventing any other workflows from running until the malformed workflow is deleted.

What version of Argo Workflows are you running?

v3.1.14

## Diagnostics

submit a simple workflow which fails, intentionally:
```yaml
# test-original.yaml
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: test-original
spec:
  entrypoint: run
  onExit: finish
  templates:
    - name: run
      script:
        image: alpine
        command: [sh]
        source: |
          echo fail on purpose
          exit 1
    - name: finish
      script:
        image: alpine
        command: [sh]
        source: |
          echo finished
```
```sh-session
$ argo submit test-original.yaml
```
Use kubectl to get that workflow:
```sh-session
kubectl get wf test-original -o yaml > test-fixed.yaml
```
Then "fix the bug" in the workflow:
```yaml
# test-fixed.yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: test-fixed
spec:
  arguments: {}
  entrypoint: run
  onExit: finish
  templates:
  - inputs: {}
    metadata: {}
    name: run
    outputs: {}
    script:
      command:
      - sh
      image: alpine
      name: ""
      resources: {}
      source: |
        echo success
  - inputs: 
    metadata: {}
    name: finish
    outputs: {}
    script:
      command:
      - sh
      image: alpine
      name: ""
      resources: {}
      source: |
        echo finished
status:
  artifactRepositoryRef:
    default: true
  conditions:
  - status: "False"
    type: PodRunning
  - status: "True"
    type: Completed
  finishedAt: "2021-11-02T17:51:25Z"
  message: Error (exit code 1)
  nodes:
    test-original-1094462238:
      displayName: test-original.onExit
      finishedAt: "2021-11-02T17:51:22Z"
      id: test-original-1094462238
      name: test-original.onExit
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: test-original/test-original-1094462238/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 5
        memory: 5
      startedAt: "2021-11-02T17:51:15Z"
      templateName: finish
      templateScope: local/test-original
      type: Pod
  phase: Failed
  progress: 2/2
  resourcesDuration:
    cpu: 10
    memory: 10
  startedAt: "2021-11-02T17:51:05Z"
```
Note that in the "fixed" workflow we've done the following to make it suitable to resubmit:
- removed all the metadata and changed to name so it doesn't conflict with `test-original`
- done a poor job of removing the status (we've only removed the key `.status.nodes.test-original` when we should have just removed the whole status)
- fixed the actual bug in the workflow (the `run` template now succeeds)

```sh-session
argo submit test-fixed.yaml
```
At this point, the `test-fixed` workflow runs and then the workflow controller pod crash loops. Only deleting the workflow `test-fixed` and then deleting the workflow controller pod allows other workflows to be picked up.

See the workflow controller logs pasted below.

Now, I realize this is pretty dumb and I wouldn't expect the workflow to run but this probably shouldn't crash the workflow-controller.

In our real world example, we had a container template in our workflow who's memory limit was set too low and it was OOMKilled. An operator used `kubectl get wf` to get the workflow, edited it locally to raise the memory limit and then submitted it. He also edited the status as described above.

What Kubernetes provider are you using?

We're not using a kubernetes provider for our real clusters. This is reproducible on kind.

What executor are you running? Docker/K8SAPI/Kubelet/PNS/Emissary

`docker` in our real clusters.
`pns` in kind.

```bash
# Logs from the workflow controller:
E1102 18:03:41.871815       1 runtime.go:78] Observed a panic: &logrus.Entry{Logger:(*logrus.Logger)(0xc000294000), Data:logrus.Fields{"fromPhase":"Failed",
"namespace":"argo", "toPhase":"Error", "workflow":"test-fixed"}, Time:time.Time{wall:0xc0587d3f73f4cba4, ext:134428595, loc:(*time.Location)(0x2cc7000)}, Lev
el:0x0, Caller:(*runtime.Frame)(nil), Message:"workflow is already fulfilled", Buffer:(*bytes.Buffer)(nil), Context:context.Context(nil), err:""} (&{0xc00029
4000 map[fromPhase:Failed namespace:argo toPhase:Error workflow:test-fixed] 2021-11-02 18:03:41.871680932 +0000 UTC m=+0.134428595 panic <nil> workflow is al
ready fulfilled <nil> <nil> })
goroutine 194 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1d94840, 0xc00068cbd0)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x2c9cab8, 0x1, 0x1)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/runtime/runtime.go:51 +0xce
panic(0x1d94840, 0xc00068cbd0)
        /usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/sirupsen/logrus.Entry.log(0xc000294000, 0xc00001b8c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/pkg/mod/github.com/sirupsen/logrus@v1.6.0/entry.go:259 +0x345
github.com/sirupsen/logrus.(*Entry).Log(0xc00068cb60, 0x0, 0xc000a1c0d0, 0x1, 0x1)
        /go/pkg/mod/github.com/sirupsen/logrus@v1.6.0/entry.go:287 +0xf0
github.com/sirupsen/logrus.(*Entry).Panic(0xc00068cb60, 0xc000a1c0d0, 0x1, 0x1)
        /go/pkg/mod/github.com/sirupsen/logrus@v1.6.0/entry.go:325 +0x55
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).markWorkflowPhase(0xc000622d10, 0x206e700, 0xc000058020, 0x1dc850c, 0x5, 0xc000a8
0270, 0xc8)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1841 +0x2bd
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).markWorkflowError(0xc000622d10, 0x206e700, 0xc000058020, 0x2021280, 0xc000179e70)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1951 +0x7c
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2(0xc000622d10, 0x206e700, 0xc000058020)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:185 +0x32c
panic(0x1d94840, 0xc00068c930)
        /usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/sirupsen/logrus.Entry.log(0xc000294000, 0xc00001b800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/pkg/mod/github.com/sirupsen/logrus@v1.6.0/entry.go:259 +0x345
github.com/sirupsen/logrus.(*Entry).Log(0xc00068c8c0, 0x0, 0xc000a1cb80, 0x1, 0x1)
        /go/pkg/mod/github.com/sirupsen/logrus@v1.6.0/entry.go:287 +0xf0
github.com/sirupsen/logrus.(*Entry).Panic(0xc00068c8c0, 0xc000a1cb80, 0x1, 0x1)
        /go/pkg/mod/github.com/sirupsen/logrus@v1.6.0/entry.go:325 +0x55
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).markWorkflowPhase(0xc000622d10, 0x206e700, 0xc000058020, 0x1dce749, 0x9, 0x0, 0x0
)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1841 +0x2bd
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).markWorkflowSuccess(...)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1943
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0xc000622d10, 0x206e700, 0xc000058020)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:436 +0x2a86
github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0xc00080c380, 0x206e700, 0xc000058020, 0x0)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:674 +0x7f0
github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc00080c380)
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:599 +0x9b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000b281b0)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000b281b0, 0x2023bc0, 0xc00001a8a0, 0x1, 0xc000138060)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000b281b0, 0x3b9aca00, 0x0, 0x1, 0xc000138060)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc000b281b0, 0x3b9aca00, 0xc000138060)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/wait/wait.go:90 +0x4d
created by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading
        /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:289 +0x48a
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by sarabala1979**
Signed-off-by: Saravanan Balasubramanian <sarabala1979@gmail.com>
closes #7139 

Don't bother creating a PR until you've done this:

* [ ] Run `make pre-commit -B` to fix codegen, lint, and commit message problems.

Create your PR as a draft.

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it
  does not need to pass.
* Once required tests have passed, you can make it "Ready for review".
* Say how how you tested your changes. If you changed the UI, attach screenshots.

Tips:

* If changes were requested, and you've made them, then dismiss the review to get it looked at again.
* Add you organization to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md) if you like.
* You can ask for help!

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -186,10 +186,12 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 	defer func() {
 		if r := recover(); r != nil {
 			woc.log.WithFields(log.Fields{"stack": string(debug.Stack()), "r": r}).Errorf("Recovered from panic")
-			if rerr, ok := r.(error); ok {
-				woc.markWorkflowError(ctx, rerr)
-			} else {
-				woc.markWorkflowError(ctx, fmt.Errorf("%v", r))
+			if !woc.wf.Status.Fulfilled() {
```
Comment (alexec) [inline workflow/controller/operator.go:189]: I think we should error the workflow,  but only if the original workflow was not fulfilled.

File: workflow/controller/controller.go [hunk 2]
```diff
@@ -767,7 +767,13 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
 		return true
 	}
 	startTime := time.Now()
-	woc.operate(ctx)
+
+	// check if workflow is already completed
+	if !woc.wf.Status.Fulfilled() {
```
Comment (sarabala1979) [inline workflow/controller/controller.go:772]: I think the right fix is here. Workflow should not operate if workflow is already fulfilled

Comment (simster7) [reply workflow/controller/controller.go:772]: I recommend doing this inside `operate`. We will still want the first deferred function to run

```go
func (woc *wfOperationCtx) operate(ctx context.Context) {
	defer func() {
		if woc.wf.Status.Fulfilled() {
			woc.killDaemonedChildren("")
		}
		woc.persistUpdates(ctx)
	}()

        // ...

	// Do not operate on Fulfilled workflows
	if woc.wf.Fulfilled() {
		return
	}

        // ...
}

Comment (sarabala1979) [reply workflow/controller/controller.go:772]: make sense. This will enable code testability.

**Review by simster7 — APPROVED** [review]

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -198,6 +198,11 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 
 	woc.log.Infof("Processing workflow")
 
+	if woc.wf.Status.Fulfilled() {
```
Comment (alexec) [inline workflow/controller/operator.go:201]: Yes, this is a better fix than I suggested, it addresses the issue where labels and status are inconsistent, presumably that is what you think the underlying cause is. 

Would it be sensible to patch the labels at this point?

**Pull request closed without merging**